sdk/javascript/client/encoder.js

```diff
@@ -0,0 +1,185 @@
+/**
+ * Copyright 2017 Intel Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * ------------------------------------------------------------------------------
+ */
+
+'use strict'
+
+const {createHash} = require('crypto')
+const Signer = require('./signer')
+const {
+  TransactionHeader,
+  Transaction,
+  BatchHeader,
+  Batch,
+  BatchList
+} = require('../protobuf')
+
+const _arrayify = itemOrItems => {
+  if (Array.isArray(itemOrItems)) return itemOrItems
+  return [itemOrItems]
+}
+
+const _makeNonce = () => {
+  const dateString = Date.now().toString(36).slice(-5)
+  const randomString = Math.floor(Math.random() * 46655).toString(36)
+  return dateString + ('00' + randomString).slice(-3)
+}
+
+/** Defines encoders which create multiple Transactions with common settings */
+class TransactionEncoder {
```

Comment:
I think I would push back on this actually. The focus of the module is on creating a Transaction, and then wrapping it in whatever packaging is necessary to get Sawtooth to accept it (or to satisfy similar use cases, like creating a Transaction to be batched elsewhere). The focus is not on creating Batches, or BatchLists, those are incidental, only necessary to package Transactions. Every method with the exception of `list` (which is my worst named, and if anything, that should probably change), is named in relation to Transactions: "create _(a transaction)_," "encode _(a transaction)_," "batch _(transactions)."

I don't know what use cases there would be for doing operations on Batches and BatchLists independent of Transactions, but I don't think this module would satisfy them. I also don't think this functionality should be broken up, as there is none of the other methods make sense without the core `create` (transaction) method.